sonar-lits-plugin/src/main/java/com/sonarsource/lits/LitsReport.java

```diff
@@ -0,0 +1,246 @@
+
+    public String ruleKey() {
+      return ruleKey;
+    }
+
+    public String componentKey() {
+      return componentKey;
+    }
+
+    public int line() {
+      return line;
+    }
+
+    public String severity() {
+      return severity;
+    }
+
+    public int key() {
+      return this.hashCode();
+    }
+  }
+
+
+  public static class ResourceReport {
+
+    Set<ReportedIssue> issues;
+
+    // moduleKey:relative/path/to/file.foo
+    String componentKey;
+
+    InputComponent inputComponent;
+
+    boolean isFile;
+
+    private Map<Integer, List<ReportedIssue>> issuesPerLine = new HashMap<>();
+
+    public ResourceReport(String componentKey, Set<ReportedIssue> issues, FileSystem fileSystem) {
+      this.componentKey = componentKey;
+      this.issues = issues;
+
+      // InputFile corresponding to this resource, it's null for new issues
+      InputComponent inputComponent = null;
+
+      // Module key, first part before colon
+      // It's required to correctly extract relative path from componentKey
+      // It's null for missing issues
+      String moduleKey = null;
+
+      for (ReportedIssue issue : this.issues) {
+        if (issue.inputComponent != null && inputComponent == null) {
+          inputComponent = issue.inputComponent;
+        }
+
+        if (issue.moduleKey != null && moduleKey == null) {
+          moduleKey = issue.moduleKey;
+        }
+
+        if (issuesPerLine.containsKey(issue.line)) {
+          issuesPerLine.get(issue.line).add(issue);
+        } else {
+          List<ReportedIssue> perLine = new ArrayList<>();
+          perLine.add(issue);
+          issuesPerLine.put(issue.line, perLine);
+        }
+      }
+
+      retrieveInputComponent(fileSystem, inputComponent, moduleKey);
+
+      this.isFile = this.inputComponent instanceof InputFile;
+
+    }
+
+    public Set<ReportedIssue> issues() {
+      return issues;
+    }
+
+    public String type() {
+      return isFile ? "FIL" : "DIR";
+    }
+
+    public String componentKey() {
+      return componentKey;
+    }
+
+    public List<ReportedIssue> issuesAtLine(int line) {
+      return issuesPerLine.getOrDefault(line, new ArrayList<>());
+    }
+
+    public boolean isDisplayableLine(@Nullable Integer lineNumber) {
+      if (lineNumber == null || lineNumber < 1) {
+        return false;
+      }
+      for (int i = lineNumber - 2; i <= lineNumber + 2; i++) {
+        if (hasIssues(i)) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    private void retrieveInputComponent(FileSystem fileSystem, @Nullable InputComponent inputComponent, @Nullable String moduleKey) {
+      if (inputComponent != null) {
+        this.inputComponent = inputComponent;
+
+      } else {
+        if (componentKey.equals(moduleKey)) {
+          throw new IllegalStateException("Module/project level issues are not supported by LITS plugin");
+        }
+
+        String relativePath = componentKey.substring(moduleKey.length() + 1);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "moduleKey" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)